README.md
[![CircleCI](https://circleci.com/gh/ThibTrip/pangres.svg?style=svg&circle-token=3e39be6b969ed02b41d259c279da0d9e63751506)](https://circleci.com/gh/ThibTrip/pangres)
[![codecov](https://codecov.io/gh/ThibTrip/pangres/branch/master/graph/badge.svg)](https://codecov.io/gh/ThibTrip/pangres)
[![PyPI version](https://img.shields.io/pypi/v/pangres)](https://img.shields.io/pypi/v/pangres)
[![Documentation](https://img.shields.io/badge/wiki-documentation-forestgreen)](https://github.com/ThibTrip/pangres/wiki)
[![Made withJupyter](https://img.shields.io/badge/Made%20with-Jupyter-orange?logo=Jupyter)](https://jupyter.org/try)

# pangres
![pangres logo](https://raw.githubusercontent.com/ThibTrip/pangres/master/logo.png)

_Thanks to [freesvg.org](https://freesvg.org/) for the logo assets_

Upsert with pandas DataFrames (<code>ON CONFLICT DO NOTHING</code> or <code>ON CONFLICT DO UPDATE</code>) for PostgreSQL, MySQL, SQlite and potentially other databases behaving like SQlite (untested) with some additional optional features (see features). Upserting can be done with **primary keys** or **unique keys**.
Pangres also handles the creation of non existing SQL tables and schemas.


# Features

1. <i>(optional)</i> Automatical column creation (when a column exists in the DataFrame but not in the SQL table)
2. <i>(optional)</i> Automatical column type alteration for columns that are empty in the SQL table (except for SQlite where alteration is limited)
3. <i>(optional)</i> Creates the table if it is missing
4. <i>(optional)</i> Creates missing schemas in Postgres (and potentially other databases that have a schema system)
5. JSON is supported (with pd.to_sql it does not work) with some exceptions (see [Gotchas and caveats](#Gotchas-and-caveats))
6. Fast (except for SQlite where some help is needed)
7. Will work even if not all columns defined in the SQL table are there
8. SQL injection safe (schema, table and column names are escaped and values are given as parameters)
9. _New in version 4.1_: **asynchronous support**. Tested using `aiosqlite` for SQlite, `asyncpg` for PostgreSQL and `aiomysql` for MySQL

# Requirements

* SQlite >= 3.24.4
* Python >= 3.6.4
* See also ./pangres/requirements.txt

# Gotchas and caveats

## All flavors
1. We can't create JSON columns automatically but we can insert JSON like objects (list, dict) in existing JSON columns.

## Postgres

1. "%", ")" and "(" in column names will most likely cause errors with PostgreSQL (this is due to psycopg2 and also affect pd.to_sql). Use the function pangres.fix_psycopg2_bad_cols to "clean" the columns in the DataFrame. You'll also have to rename columns in the SQL table accordingly (if the table already exists).
2. Even though we only do data type alteration on empty columns, since we don't want to lose column information (e.g. constraints) we use true column alteration (instead of drop+create) so the old data type must be castable to the new data type. Postgres seems a bit restrictive in this regard even when the columns are empty (e.g. BOOLEAN to TIMESTAMP is impossible).

## SQlite
1. **SQlite must be version 3.24.4 or higher**! UPSERT syntax did not exist before. 
2. Column type alteration is not possible for SQlite.
3. SQlite inserts can be at worst 5 times slower than pd.to_sql for some reasons. If you can help please contact me!
4. Inserts with 1000 columns (or 32767 columns for SQlite >= 3.32.0) or more are not supported because one could not even insert one row without exceeding the max number of parameters per queries. One way to fix this would inserting the columns progressively but this seems quite tricky. If you know a better way please contact me.

## MySQL

1. MySQL will often change the order of the primary keys in the SQL table when using INSERT... ON CONFLICT.. DO NOTHING/UPDATE. This seems to be the expected behavior so nothing we can do about it but please mind that!
2. You may need to provide SQL dtypes e.g. if you have a primary key with text you will need to provide a character length (e.g. VARCHAR(50)) because MySQL does not support indices/primary keys with flexible text length. pd.to_sql has the same issue.


# Notes

This is a library I was using in production in private with very good results and decided to publish.

Ideally such features will be integrated into pandas since there is already a [PR on the way](https://github.com/pandas-dev/pandas/pull/29636) and I would like to give the option to add columns via another PR.

There is also [pandabase](https://github.com/notsambeck/pandabase) which does almost the same thing (plus lots of extra features) but my implementation is different.
Btw big thanks to pandabase and the sql part of pandas which helped a lot.

# Installation
```
pip install pangres
```
Additionally depending on which database you want to work with you will need to install the corresponding library (note that SQlite is included in the standard library):

* Postgres
```
pip install psycopg2
```

* MySQL
```
pip install pymysql
```

* Postgres (asynchronous)
```
pip install asyncpg
```

* MySQL (asynchronous)
```
pip install aiomysql
```

* SQLite (asynchronous)
```
pip install aiosqlite
```

# Usage

Head over to [pangres' wiki](https://github.com/ThibTrip/pangres/wiki)!

A big part of the wiki was generated with my library [**npdoc_to_md**](https://github.com/ThibTrip/npdoc_to_md).

# Contributing

Pull requests/issues are welcome.

# Development

I develop the library inside of **Jupyter Lab** using the [**jupytext**](https://github.com/mwouts/jupytext) extension.

I recommend using this extension for the best experience.
It will split code blocks within modules in notebook cells and will allow **interactive development**.

If you wish you can also use the provided **conda environment** (see `environment.yml` file) inside of Jupyter Lab/Notebook
thanks to [**nb_conda_kernels**](https://github.com/Anaconda-Platform/nb_conda_kernels).

# Testing

You can test one or multiple of the following SQL flavors (you will of course need a live database for this): PostgreSQL, SQlite or MySQL.

NOTE: in one of the tests of `pangres` we will try to drop and then create a PostgreSQL schema called `pangres_create_schema_test`. If the schema existed and was not empty an error will be raised.

Clone pangres then set your curent working directory to the root of the cloned repository folder. Then use the commands below. You will have to replace the following variables in those commands:
* `SQLITE_CONNECTION_STRING`: replace with a SQlite sqlalchemy connection string (e.g. "sqlite:///test.db")
* `ASYNC_SQLITE_CONNECTION_STRING`: replace with an asynchronous SQlite sqlalchemy connection string (e.g. "sqlite+aiosqlite:///test.db")
* `POSTGRES_CONNECTION_STRING`: replace with a Postgres sqlalchemy connection string (e.g. "postgres:///user:password@localhost:5432/database"). Specifying schema is optional for postgres (will default to public)
* `ASYNC_POSTGRES_CONNECTION_STRING`: replace with an asynchronous Postgres sqlalchemy connection string (e.g. "postgres+asyncpg:///user:password@localhost:5432/database"). Specifying schema is optional for postgres (will default to public)
* `MYSQL_CONNECTION_STRING`: replace with a MySQL sqlalchemy connection string (e.g. "mysql+pymysql:///user:password@localhost:3306/database")
* `ASYNC_MYSQL_CONNECTION_STRING`: replace with an asynchronous MySQL sqlalchemy connection string (e.g. "mysql+aiomysql:///user:password@localhost:3306/database")
* `PG_SCHEMA` (optional): schema for postgres (defaults to public)

```shell
# 1. Create and activate the build environment
conda env create -f environment.yml
conda activate pangres-dev
# 2. Install pangres in editable mode (changes are reflected upon reimporting)
pip install -e .
# 3. Run pytest
# -s prints stdout
# -v prints test parameters
# --cov=./pangres shows coverage only for pangres
# --doctest-modules tests with doctest in all modules
# --benchmark-XXX : these are options for benchmarks tests (see https://pytest-benchmark.readthedocs.io/en/latest/usage.html)
pytest -s -v pangres --cov=pangres --doctest-modules --async_sqlite_conn=$ASYNC_SQLITE_CONNECTION_STRING --sqlite_conn=$SQLITE_CONNECTION_STRING --async_pg_conn=$ASYNC_POSTGRES_CONNECTION_STRING --pg_conn=$POSTGRES_CONNECTION_STRING --async_mysql_conn=$ASYNC_MYSQL_CONNECTION_STRING --mysql_conn=$MYSQL_CONNECTION_STRING --pg_schema=tests --benchmark-group-by=func,param:engine,param:nb_rows --benchmark-columns=min,max,mean,rounds --benchmark-sort=name --benchmark-name=short
```

Additionally, the following flags could be of interest for you:
* `-x` for stopping at the first failure
* `--benchmark-only` for only testing benchmarks
* `--benchmark-skip` for skipping benchmarks


CHANGELOG.md
# Changelog

## [v4.1.2](https://github.com/ThibTrip/pangres/releases/tag/v4.1.2) (2022-07-29)


**Release Notes**


___
**_Bug Fixes_**


* **`pangres` was not running on Python 3.10** because I never added `packaging` in `requirements.txt`. I fixed it by using `pkg_resources` from the standard library instead to avoid the additional dependency (see commit 89d3679)

  Note: the tests were running fine because we use `pytest` which uses `packaging` (so I never saw the missing dependency)
___



[Full Changelog](https://github.com/ThibTrip/pangres/compare/v4.1.1...v4.1.2)

**Closed issues:**

- Cannot insert into column "id" [\#60](https://github.com/ThibTrip/pangres/issues/60)

## [v4.1.1](https://github.com/ThibTrip/pangres/releases/tag/v4.1.1) (2022-03-13)


**Release Notes**


___
**_Bug Fixes_**

* fixed bug where I used a synchronous method instead of its asynchronous variant (`UpsertQuery.execute` instead of `UpsertQuery.aexecute` in `pangres.aupsert`). This has no repercussions for the end user

**_Documentation_**

* fix illogic code in example for `pangres.aupsert` (using `engine` instead of `connection` in contexts) and `commit` which I had forgotten!
* added changelog

**_Testing_**

* overhaul of the tests. asynchronous and synchronous tests have been separated
* module `test_upsert_end_to_end` has been renamed to `test_core`
___



[Full Changelog](https://github.com/ThibTrip/pangres/compare/v4.1...v4.1.1)

**Fixed bugs:**

- Fix bad strategy of temporarily replacing async engines with synchronous engines in tests [\#57](https://github.com/ThibTrip/pangres/issues/57)

**Closed issues:**

- Changelog \(releases\) [\#58](https://github.com/ThibTrip/pangres/issues/58)

**Merged pull requests:**

- \[TESTS\] Tests overhaul [\#59](https://github.com/ThibTrip/pangres/pull/59) ([ThibTrip](https://github.com/ThibTrip))

## [v4.1](https://github.com/ThibTrip/pangres/releases/tag/v4.1) (2022-01-21)


**Release Notes**


___
**_New Features_**

* Added async support with function `pangres.aupsert` :rocket: ! Tested using `aiosqlite` for SQlite, `asyncpg` for PostgreSQL and `aiomysql` for MySQL. See documentation in dedicated wiki [page](https://github.com/ThibTrip/pangres/wiki/Aupsert)
___



[Full Changelog](https://github.com/ThibTrip/pangres/compare/v4.0.2...v4.1)

**Implemented enhancements:**

- \[FEAT\] Add `pangres.aupsert` \(async variant\) [\#55](https://github.com/ThibTrip/pangres/pull/55) ([ThibTrip](https://github.com/ThibTrip))

**Closed issues:**

- Add async support? [\#47](https://github.com/ThibTrip/pangres/issues/47)

## [v4.0.2](https://github.com/ThibTrip/pangres/releases/tag/v4.0.2) (2022-01-17)


**Release Notes**


___
This patches an important bug with MySQL. We recommend that all users upgrade to this version.

**_Bug Fixes_**

* Fixed bug where tables in MySQL where created with auto increment on the primary key (see #56)
___



[Full Changelog](https://github.com/ThibTrip/pangres/compare/v4.0.1...v4.0.2)

**Closed issues:**

- MySQL creating undesired auto incrementing primary keys [\#56](https://github.com/ThibTrip/pangres/issues/56)
- Get rid of the LooseVersion warning [\#54](https://github.com/ThibTrip/pangres/issues/54)

## [v4.0.1](https://github.com/ThibTrip/pangres/releases/tag/v4.0.1) (2022-01-13)


**Release Notes**


___
**_Bug Fixes_**

* removed warning due to deprecated code when checking versions of other libraries in Python >= 3.10 (see issue  #54)
___



[Full Changelog](https://github.com/ThibTrip/pangres/compare/v4.0...v4.0.1)

## [v4.0](https://github.com/ThibTrip/pangres/releases/tag/v4.0) (2022-01-12)


**Release Notes**


___
There have been important changes in the main function **`pangres.upsert`**:

**_Breaking changes_**


1. The first argument **`engine`** has been be **renamed** to **`con`** and now accepts engines and connections
2. The argument **`chunksize`** now **defaults** to **`None`**. Like in `pandas.DataFrame.to_sql` we will attempt to insert all rows by default. Previously the default was `10000` rows.
3. There will be **no more automatic adjustments to the `chunksize`** the user passes in `pangres.upsert` even if we can predict that it will raise an Exception due to database limitations.

E.g. inserting 100000 rows at once in a SQlite database with `pangres` will necessarily raise an Exception down the line because we need to pass NUMBER_OF_ROWS * NUMBER_OF_COLUMNS parameters and the maximum of parameters allowed in SQLite is 32766 (for version >= 3.32.0, otherwise it is 999).

I have made a new utility function [**`pangres.adjust_chunksize`**](https://github.com/ThibTrip/pangres/wiki/Adjust-Chunksize) that you can use before calling **`pangres.upsert`** if you want to make sure the `chunksize` is not too big.

**_New Features_**


Now that `pangres.upsert` accept connections objects this will give you more control when using it when it comes to connections and transactions.

See [transaction control demo notebook](https://github.com/ThibTrip/pangres/blob/master/demos/transaction_control.ipynb).
___



[Full Changelog](https://github.com/ThibTrip/pangres/compare/v3.1...v4.0)

**Closed issues:**

- Set default `chunksize` to None [\#51](https://github.com/ThibTrip/pangres/issues/51)
- Option to Disable commit in execute\(\) [\#44](https://github.com/ThibTrip/pangres/issues/44)

**Merged pull requests:**

- \[CORE\] Directly move to v4 [\#53](https://github.com/ThibTrip/pangres/pull/53) ([ThibTrip](https://github.com/ThibTrip))
- \[CORE\] Default `chunksize` to `None` [\#52](https://github.com/ThibTrip/pangres/pull/52) ([ThibTrip](https://github.com/ThibTrip))
- \[FEAT\] Add transaction and connection control [\#50](https://github.com/ThibTrip/pangres/pull/50) ([ThibTrip](https://github.com/ThibTrip))

## [v3.1](https://github.com/ThibTrip/pangres/releases/tag/v3.1) (2022-01-07)


**Release Notes**


___
**_Bug Fixes_**


pangres/utils.py
# +
import logging
import pandas as pd
from math import floor
from sqlalchemy.engine.base import Connectable

# local imports
from pangres.helpers import _sqlite_gt3_32_0, validate_chunksize_param
from pangres.logger import log
from pangres.exceptions import (DuplicateLabelsException,
                                TooManyColumnsForUpsertException,
                                UnnamedIndexLevelsException)


# -

# # Function to fix bad column names for psycopg2

def fix_psycopg2_bad_cols(df:pd.DataFrame, replacements:dict={'%':'', '(':'', ')':''}) -> pd.DataFrame:
    """
    Replaces '%', '(' and ')' (characters that won't play nicely or even
    at all with psycopg2) in column and index names in a deep copy of df.
    This is a workaround for the unresolved issue
    described here: https://github.com/psycopg/psycopg2/issues/167

    **IMPORTANT**:
    You will need to apply the same changes in the database as
    well if the SQL table already exists for a given DataFrame.
    Otherwise you will for instance end up with a column
    "total_%" and "total_" in your SQL table.

    Parameters
    ----------
    df : pd.DataFrame
    replacements : dict {'%':str, '(':str, ')':str}, default {'%':'', '(':'', ')':''}
        The keys '%', '(' and ')' are mandatory.
        There cannot be any extra keys.

    Raises
    ------
    pangres.exceptions.UnnamedIndexLevelsException
        When you pass a df where not all index levels are named
    pangres.exceptions.DuplicateLabelsException
        When you pass a df with duplicated labels accross index/columns
        or when after cleaning we end up with duplicated labels
        e.g. "test(" and "test)" would by default both be renamed to "test"
    TypeError
        When `replacements` is not of the expected type or has wrong keys
        or has non string values

    Returns
    -------
    new_df : pd.DataFrame

    Examples
    --------
    * fix bad col/index names with default replacements (empty string for '(', ')' and '%')
    >>> from pangres import fix_psycopg2_bad_cols
    >>> import pandas as pd
    >>> df = pd.DataFrame({'test()':[0],
    ...                    'foo()%':[0]}).set_index('test()')
    >>> print(df.to_markdown())
    |   test() |   foo()% |
    |---------:|---------:|
    |        0 |        0 |

    >>> df_fixed = fix_psycopg2_bad_cols(df)
    >>> print(df_fixed.to_markdown())
    |   test |   foo |
    |-------:|------:|
    |      0 |     0 |

    * fix bad col/index names with custom replacements - you MUST provide replacements for '(', ')' and '%'!
    >>> import pandas as pd
    >>> df = pd.DataFrame({'test()':[0],
    ...                    'foo()%':[0]}).set_index('test()')
    >>> print(df.to_markdown())
    |   test() |   foo()% |
    |---------:|---------:|
    |        0 |        0 |

    >>> df_fixed = fix_psycopg2_bad_cols(df, replacements={'%':'percent', '(':'', ')':''})
    >>> print(df_fixed.to_markdown())
    |   test |   foopercent |
    |-------:|-------------:|
    |      0 |            0 |
    """
    # verify all index levels are named
    index_names = list(df.index.names)
    if any([ix_name is None for ix_name in index_names]):
        raise UnnamedIndexLevelsException("All index levels must be named!")

    # verify duplicated columns
    fields = list(df.index.names) + df.columns.tolist()
    if len(set(fields)) != len(fields):
        duplicates = [c for c in fields if fields.count(c) > 1]
        raise DuplicateLabelsException("There cannot be duplicated names amongst "
                                       f"index levels and/or columns! Duplicates found: {duplicates}")
    # verify replacements arg
    expected_keys = ('%', '(', ')')
    if ((not isinstance(replacements, dict)) or
        (set(replacements.keys()) - set(expected_keys) != set()) or
        (len(replacements) != len(expected_keys))):
        raise TypeError(f'replacements must be a dict containing the following keys (and none other): {expected_keys}')
    if not all((isinstance(v, str) for v in replacements.values())):
        raise TypeError('The values of replacements must all be strings')

    # replace bad col names
    translator = {ord(k):v for k, v in replacements.items()}
    new_df = df.copy(deep=True)
    renamer = lambda col: col.translate(translator) if isinstance(col, str) else col
    new_df = new_df.rename(columns=renamer).rename_axis(index=renamer)

    # check columns are unique after renaming
    fields = list(new_df.index.names) + new_df.columns.tolist()
    if len(set(fields)) != len(fields):
        duplicates = [c for c in fields if fields.count(c) > 1]
        raise DuplicateLabelsException("Columns/index are not unique after renaming! "
                                       f"Duplicates found: {duplicates}")

    # compare columns (and index)
    before = df.reset_index().columns.tolist()
    after = new_df.reset_index().columns.tolist()
    for i, j in zip(before, after):
        if (isinstance(i, str) and isinstance(j, str) and i != j):
            log(f'Renamed column/index "{i}" to "{j}s"')
    return new_df


# ## Function to adjust the size of chunks to upsert (depending on a DataFrame's shape and what a database allows for SQL parameters)

def adjust_chunksize(con:Connectable, df:pd.DataFrame, chunksize:int):
    """
    Checks if given `chunksize` is appropriate for upserting rows in given database using
    given DataFrame.
    The `chunksize` represents the number of rows you would like to upsert in a chunk when
    using the `pangres.upsert` function.

    Some databases have limitations on the number of SQL parameters and we need one parameter
    per value for upserting data.
    E.g. a DataFrame of 5 **columns+index levels** and 4 rows will require 5*4=20 SQL parameters.

    This function will check the database type (e.g. SQlite) and the number of **columns+index levels**
    to determine if the `chunksize` does not exceed limits and propose a lower one if it does.
    Otherwise the same `chunksize` that you gave as input is returned.

    This function currently takes into account max parameters limitations for the following cases:
    * sqlite (32766 max for version >= 3.22.0 otherwise 999)
    * asyncpg (32767 max)

    If you know about more parameter limitations relevant for this library (PostgreSQL, MySQL, SQlite
    or other databases I have not tested with this library that you managed to have working),
    please contact me.

    Parameters
    ----------
    con
        sqlalchemy Engine or Connection
    df
        DataFrame you would wish to upsert
    chunksize
        Size of chunks you would wish to use for upserting (represents the number of rows
        in each chunk)

    Raises
    ------
    TooManyColumnsForUpsertException
        When a DataFrame has more columns+index levels than the maximum number of allowed SQL variables
        for a SQL query for given database.
        In such a case even inserting row by row would not be possible because we would already
        have too many variables.
        For more information you can for instance google "SQLITE_MAX_VARIABLE_NUMBER".

    Examples
    --------
    >>> from sqlalchemy import create_engine
    >>>
    >>> # config (this assumes you have SQlite version >= 3.22.0)
    >>> engine = create_engine("sqlite://")
    >>>
    >>> # some df we want to upsert
    >>> df = pd.DataFrame({'name':['Albert']}).rename_axis(index='profileid')
    >>> print(df.to_markdown())
    |   profileid | name   |
    |------------:|:-------|
    |           0 | Albert |

    {{python}}
    >>> # adjust chunksize: 100,000 is too big of a chunksize in general for given database
    >>> # SQlite only allows 32766 parameters (values) at once maximum in a query
    >>> # since we have two columns (technically 1 column + 1 index level)
    >>> # we can only upsert in chunks of FLOOR(32766/2) rows maximum which is 16383
    >>> adjust_chunksize(con=engine, df=df, chunksize=100_000)
    16383
    """
    validate_chunksize_param(chunksize=chunksize)

    # get maximum number of parameters depending on the database
    dialect = con.dialect.dialect_description
    if 'sqlite' in dialect:
        maximum = 32766 if _sqlite_gt3_32_0() else 999
    elif 'asyncpg' in dialect:
        maximum = 32767
    else:
        maximum = None

    # simple case we can solve early
    if maximum is None:
        return chunksize

    # adjust chunksize
    new_chunksize = floor(maximum / (len(df.columns) + df.index.nlevels))
    if new_chunksize < 1:
        raise TooManyColumnsForUpsertException('The df has more columns+index levels than the maxmimum number of allowed parameters '
                                               'for given database for a query (we could not even upsert row by row).')
    if chunksize > new_chunksize:
        log(f'Reduced chunksize from {chunksize} to {new_chunksize} due '
            f'to max variable restriction of given dialect (max {maximum} for dialect {dialect}).',
            level=logging.INFO)
        chunksize = new_chunksize
    return chunksize
